### PR TITLE
fix proposal deposit attack

### PR DESCRIPTION
**Pull request by CJentzsch**
fixes https://github.com/slockit/DAO/issues/65

### CODE REVIEW

File: DAO.sol [hunk 1]
```diff
@@ -548,7 +556,7 @@ contract DAO is DAOInterface, Token, TokenSale {
             // p.proposalDeposit should be zero here
             if (this.balance < p.proposalDeposit)
```
Comment (pirapira) [inline DAO.sol]: This overflow check should use `sumOfProposalDeposits` instead of `p.proposalDeposit`.

**Comment by colm** [comment]
I believe this opens the door to a 51% attacker preventing splits. With the following actions

1. The attacker creates an ordinary proposal with a deposit that is equal to the current `this.balance`
2. The attacker prevents a quorum being met on this proposal. 
3. When anyone tries to split their new ETH balance will be zero 
`p.splitData[0].splitBalance = this.balance - sumOfProposalDeposits;`
4. The attacker can create a proposal to send the proposal deposit to an address they choose. If the deposit is moved tis way `sumOfProposalDeposits` will never be decremented. Effectively preventing all splits for a temporary deposit.

**Comment by LefterisJP** [comment]
But @colm would not step (1) effectively double `this.balance`?

**Comment by colm** [comment]
Oh yes it would. So let me adjust the attack a little. 

1. creates proposal which huge deposit
2. prevent quoram 
3. Create a proposal to send the amount of the `proposalDeposit` to its service provider. This will decrease `this.balance` but `sumOfProposalDeposits` would remain unchanged. 
4. Splits are prevented. 

If the attacker has trouble getting a quoram for such a large amount of ETH. They can do this 1,000 times with a proposal deposit 1,000 times smaller.

**Comment by LefterisJP** [comment]
> Create a proposal to send the amount of the proposalDeposit to its service provider. This will decrease this.balance but sumOfProposalDeposits would remain unchanged.

No it would not. It would only decrease `this.balance` once the proposal is executed. So once the debating period is over and he calls `executeProposal()`.

**Comment by CJentzsch** [comment]
@colm you can only spend so much ether, that the proposalDeposit remains
https://github.com/slockit/DAO/pull/73/files#diff-b30f9d3be2a832949bdd72e23956c130R498

And, after the proposal debating period ends, everyon can call `executeProposal` and the umOfProposalDeposits` will be decreased, independent of quourum:
https://github.com/slockit/DAO/pull/73/files#diff-b30f9d3be2a832949bdd72e23956c130R522

**Comment by colm** [comment]
Oh yes thats fine then :)

**Comment by colm** [comment]
Should we subtract `sumOfProposalDeposits` [here](https://github.com/slockit/DAO/blob/master/DAO.sol#L693) and [here](https://github.com/slockit/DAO/blob/master/DAO.sol#L722) 

To protect against setting the proposal deposit greater than 10% of the eth balance and abuse of the `minQuorum` formula.

**Comment by LefterisJP** [comment]
@colm yes I believe you are right. Let me handle that.

**Comment by pirapira** [comment]
It might be simpler to store the deposits in a separate managed account.  Then no subtraction is necessary.

**Comment by LefterisJP** [comment]
@pirapira Not a bad idea. Adding yet another ManagedAccount could potentially reduce code readability though.

For the moment I added a function to abstract this whole thing away and avoid code repetition.
@CJentzsch what do you think?

**Comment by CJentzsch** [comment]
@pirapira Seperate account is good. But then the DAO constantly would need to grab the ether from there for proposal which have not reached the quorum. and I actually think it would add complexity.

**Pull request merged**